test/plugin_testsetup.jl

```diff
@@ -69,5 +69,143 @@ function inlining_handler(meta::InlineStateMeta, interp::GPUInterpreter, @nospec
+        callinfo = DeferredCallInfo(MockEnzymeMeta(), call.rt, call.info)
+
+        # Real Enzyme must compute `rt` and `exct` according to enzyme semantics
+        # and likely perform a unwrapping of fargs...
+        rt = call.rt
+
+        # TODO: Edges? Effects?
+        @static if VERSION < v"1.11.0-"
+            # Can't use call.effects since otherwise this call might be just replaced with rt
+            return CallMeta(rt, Effects(), AutodiffCallInfo(rt, callinfo))
+        else
+            return CallMeta(rt, call.exct, Effects(), AutodiffCallInfo(rt, callinfo))
+        end
+    end
+
+    return nothing
+end
+
+import Core.Compiler: insert_node!, NewInstruction, ReturnNode, Instruction, InliningState, Signature
+
+# We really need a Compiler stdlib
+Base.getindex(ir::IRCode, i) = Core.Compiler.getindex(ir, i)
+Base.setindex!(inst::Instruction, val, i) = Core.Compiler.setindex!(inst, val, i)
+
+const FlagType = VERSION >= v"1.11.0-" ? UInt32 : UInt8
+function Core.Compiler.handle_call!(todo::Vector{Pair{Int,Any}}, ir::IRCode, stmt_idx::Int,
+                         stmt::Expr, info::AutodiffCallInfo, flag::FlagType,
+                         sig::Signature, state::InliningState)
+
+    # Goal:
+    # The IR we want to inline here is:
+    # unpack the args ..
+    # ptr = gpuc.deferred(MockEnzymeMeta(), f, primal_args...) 
+    # ret = ccall("extern __autodiff", llvmcall, RT, Tuple{Ptr{Cvoid, args...}}, ptr, adjoint_args...)
+
+    # 0. Obtain primal mi from DeferredCallInfo
+    # TODO: remove this code duplication
+    deferred_info = info.info
+    minfo = deferred_info.info
+    results = minfo.results
+    if length(results.matches) != 1
+        return nothing
+    end
+    match = only(results.matches)
+
+    # lookup the target mi with correct edge tracking
+    # TODO: Effects?
+    case = Core.Compiler.compileable_specialization(
+        match, Core.Compiler.Effects(), Core.Compiler.InliningEdgeTracker(state), info)
+    @assert case isa Core.Compiler.InvokeCase
+    @assert stmt.head === :call
+
+    # Now create the IR we want to inline
+    ir = Core.Compiler.IRCode() # contains a placeholder
+    args = [Core.Compiler.Argument(i) for i in 2:length(stmt.args)] # f, args...
+    idx = 0
+
+    # 0. Enzyme proper: Desugar args 
+    primal_args = args
+    primal_argtypes = match.spec_types.parameters[2:end]
+
+    adjoint_rt = info.rt
+    adjoint_args = args # TODO
+    adjoint_argtypes = primal_argtypes
+    
+    # 1: Since Julia's inliner goes bottom up we need to pretend that we inlined the deferred call
+    expr = Expr(:foreigncall, 
+        "extern gpuc.lookup",
+        Ptr{Cvoid},
+        Core.svec(#=meta=# Any, #=mi=# Any, #=f=# Any, primal_argtypes...), # Must use Any for MethodInstance or ftype
+        0,
+        QuoteNode(:llvmcall),
+        deferred_info.meta,
+        case.invoke,
+        primal_args...
+    )
+    ptr = insert_node!(ir, (idx += 1), NewInstruction(expr, Ptr{Cvoid}))
```

Comment:
@aviatesk does this look correct?

Reply:
`handle_call!` isn’t meant to be overloaded, so I think this approach is preferred:
```julia
function Core.Compiler.src_inlining_policy(interp::GPUCompiler#=or EnzymeInterpreter?=#,
    @nospecialize(src), info::AutodiffCallInfo, stmt_flag::UInt32)
    # Goal:
    # The IR we want to return here is:
    # unpack the args ..
    # ptr = gpuc.deferred(MockEnzymeMeta(), f, primal_args...) 
    # ret = ccall("extern __autodiff", llvmcall, RT, Tuple{Ptr{Cvoid, args...}}, ptr, adjoint_args...)
    ir = Core.Compiler.IRCode() # contains a placeholder
    ...
    return ir
end
```
By overloading `src_inlining_policy` (or `inlining_policy` in older versions), we can apply this custom inlining to const-propped call sites and semi-concrete interpreted call sites as well.

Reply:
Haha, then I have misunderstood the comment in:

https://github.com/JuliaLang/julia/blob/b9d9b69165493f6fc03870d975be05c67f14a30b/base/compiler/ssair/inlining.jl#L1668-L1669

Reply:
Ah, I just realized that overloading `src_inlining_policy` wouldn't be enough.. We need to overload `retrieve_ir_for_inlining` too, but it doesn't take `info::CallInfo`, so maybe we need to tweak the interface..

But I believe this approach (overloading `inlining_policy`) works at least for pre-1.11.

> Haha, then I have misunderstood the comment in:
> 
> https://github.com/JuliaLang/julia/blob/b9d9b69165493f6fc03870d975be05c67f14a30b/base/compiler/ssair/inlining.jl#L1668-L1669

It seems like I’ve ended up betraying my past self.

Reply:
Yeah I think you once told me to extend handle_call